helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java

```diff
@@ -172,11 +173,15 @@
   private boolean _inMaintenanceMode;
 
   /**
-   * The timer that can periodically run the rebalancing pipeline. The timer will start if there is
-   * one resource group has the config to use the timer.
+   * The executors that can periodically run the rebalancing pipeline. A
+   * SingleThreadScheduledExecutor will start if there is one resource group has the config to do
+   * periodically rebalance.
    */
-  Timer _periodicalRebalanceTimer = null;
+  private static final ScheduledExecutorService _periodicalRebalanceExecutor =
+      Executors.newSingleThreadScheduledExecutor();
+  private ScheduledFuture _periodicRebalancerFutureTasks = null;
   long _timerPeriod = Long.MAX_VALUE;
+  private final Object _lock = new Object();
```

Comment:
Just lock the _periodicalRebalanceExecutor?
&
Have you tried the atomic reference solution that we discussed? I think that one is easier.

Reply:
TFTR. Updated.